kernel/src/fs/cgroupfs/controller/mod.rs

```diff
@@ -0,0 +1,413 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{string::String, sync::Arc, vec::Vec};
+
+use aster_systree::{Error, Result, SysAttrSet, SysAttrSetBuilder};
+use bitflags::bitflags;
+use ostd::{
+    mm::{VmReader, VmWriter},
+    sync::{Mutex, MutexGuard, Rcu},
+    task::{atomic_mode::AsAtomicModeGuard, disable_preempt},
+};
+
+use crate::fs::cgroupfs::{
+    controller::{cpuset::CpuSetController, memory::MemoryController, pids::PidsController},
+    systree_node::CgroupSysNode,
+    CgroupNode,
+};
+
+mod cpuset;
+mod memory;
+mod pids;
+
+/// A trait to abstract all individual cgroup controllers.
+trait SubControl {
+    fn attr_set(&self) -> &SysAttrSet;
+
+    fn read_attr_at(
+        &self,
+        name: &str,
+        offset: usize,
+        writer: &mut VmWriter,
+        cgroup_node: &dyn CgroupSysNode,
+    ) -> Result<usize>;
+
+    fn write_attr(
+        &self,
+        name: &str,
+        reader: &mut VmReader,
+        cgroup_node: &dyn CgroupSysNode,
+    ) -> Result<usize>;
+}
+
+/// Defines the static properties and behaviors of a specific cgroup sub-controller.
+trait SubControlStatic: SubControl + Sized + 'static {
+    /// Creates a new instance of the sub-controller.
+    fn new(is_root: bool) -> Self;
+
+    /// Returns the `ControllerType` enum variant corresponding to this sub-controller.
+    fn controller_type() -> ControllerType;
+
+    /// Obtains the specific `SubController` field from the main `Controller` struct.
+    fn obtain_sub_controller_field(controller: &Controller) -> Arc<SubController<Self>>;
+}
+
+/// The type of sub-controller in the cgroup subsystem.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum ControllerType {
+    Memory,
+    CpuSet,
+    Pids,
+}
+
+impl ControllerType {
+    const ALL: [Self; 3] = [Self::Memory, Self::CpuSet, Self::Pids];
+}
+
+impl TryFrom<&str> for ControllerType {
+    type Error = aster_systree::Error;
+
+    fn try_from(value: &str) -> Result<Self> {
+        match value {
+            "memory" => Ok(ControllerType::Memory),
+            "cpuset" => Ok(ControllerType::CpuSet),
+            "pids" => Ok(ControllerType::Pids),
+            _ => Err(Error::NotFound),
+        }
+    }
+}
+
+bitflags! {
+    /// Bitflags representing active/deactive sub-control state.
+    pub(super) struct SubCtrlState: u8 {
```

Comment:
I think it is better to name this as `SubCtrlSet`.
```rust
/// A set of sub-controller types, represented as bitflags.
struct SubCtrlSet {}
```

So we can have:
```rust
pub(super) struct Controller {
    /// A set of types of active sub-controllers.
    active_set: Mutex<SubCtrlSet>,
```
and
```rust
fn read_subtree_control_from_reader(
    reader: &mut VmReader,
) -> Result<(SubCtrlSet, SubCtrlSet, usize)> {
```

It perhaps makes more sense.